administrator/components/com_jedchecker/libraries/rules/xmlmanifest.php

```diff
@@ -0,0 +1,404 @@
+
+	/**
+	 * The description of this rule.
+	 *
+	 * @var    string
+	 */
+	protected $description = 'COM_JEDCHECKER_MANIFEST_DESC';
+
+	/**
+	 * List of errors.
+	 *
+	 * @var    string[]
+	 */
+	protected $errors;
+
+	/**
+	 * List of warnings.
+	 *
+	 * @var    string[]
+	 */
+	protected $warnings;
+
+	/**
+	 * List of infos.
+	 *
+	 * @var    string[]
+	 */
+	protected $infos;
+
+	/**
+	 * Rules for XML nodes
+	 *   ? - single, optional
+	 *   = - single, required, warning if missed
+	 *   ! - single, required, error if missed
+	 *   * - multiple, optional
+	 * @var array
+	 */
+	protected $DTDNodeRules;
+
+	/**
+	 * Rules for attributes
+	 *   (list of allowed attributes)
+	 * @var array
+	 */
+	protected $DTDAttrRules;
+
+	/**
+	 * List of extension types
+	 *
+	 * @var string[]
+	 */
+	protected $joomlaTypes = array(
+		'component', 'file', 'language', 'library',
+		'module', 'package', 'plugin', 'template'
+	);
+
+	/**
+	 * List of JED extension types
+	 *
+	 * @var string[]
+	 */
+	protected $jedTypes = array(
+		'component', 'module', 'package', 'plugin'
+	);
+
+	/**
+	 * Initiates the search and check
+	 *
+	 * @return    void
+	 */
+	public function check()
+	{
+		// Find all XML files of the extension
+		$files = JFolder::files($this->basedir, '\.xml$', true, true);
+
+		// Iterate through all the xml files
+		foreach ($files as $file)
+		{
+			// Try to check the file
+			$this->find($file);
+		}
+	}
+
+	/**
+	 * Reads a file and validate XML manifest
+	 *
+	 * @param   string  $file  - The path to the file
+	 *
+	 * @return boolean True if the manifest file was found, otherwise False.
+	 */
+	protected function find($file)
+	{
+		$xml = simplexml_load_file($file);
+
+		// Failed to parse the xml file.
+		// Assume that this is not a extension manifest
+		if (!$xml)
+		{
+			return false;
+		}
+
+		// Check if this is an extension manifest
+		if ($xml->getName() !== 'extension')
+		{
+			return false;
+		}
+
+		// Check extension type
+		$type = (string) $xml['type'];
+
+		if (!in_array($type, $this->joomlaTypes, true))
+		{
+			$this->report->addError($file, JText::sprintf('COM_JEDCHECKER_MANIFEST_UNKNOWN_TYPE', $type));
+
+			return true;
+		}
+
+		// JED allows components, modules, plugins, and packages (as a container) only
+		if (!in_array($type, $this->jedTypes, true))
+		{
+			$this->report->addError($file, JText::sprintf('COM_JEDCHECKER_MANIFEST_TYPE_NOT_ACCEPTED', $type));
+		}
+
+		// Load DTD-like data for this extension type
+		$jsonFilename = __DIR__ . '/xmlmanifest/dtd_' . $type . '.json';
+
+		if (!is_file($jsonFilename))
+		{
+			return true;
+		}
+
+		// Warn if method="upgrade" attribute is not found
+		if ((string) $xml['method'] !== 'upgrade')
+		{
+			$this->report->addWarning($file, JText::_('COM_JEDCHECKER_MANIFEST_MISSED_METHOD_UPGRADE'));
+		}
+
+		// Check 'client' attribute is "site" or "administrator" (for module/template only)
+		if ($type === 'module' || $type === 'template')
```

Comment:
by design template is not checked because of https://github.com/joomla-extensions/jedchecker/pull/81/files#diff-56791c24510fc7a98c5b0e18f369272af0076e69fba99be7aceff37c015b737cR157 so doing any checks for any other types then those allowed seems pointless.

Reply:
1) Template may be a part of a package (e.g. an AMP extension consisting of a plugin and a template),
2) It costs almost nothing.
PS. And it would be nice to have Joomla templates directory along with JED.

Reply:
1. My point is that at this [above check](https://github.com/joomla-extensions/jedchecker/blob/c60b397af7b4fe36b25ff84dcabfbceb522d3c3a/administrator/components/com_jedchecker/libraries/rules/xmlmanifest.php#L157) we must break out with a return since if the package is not excepted in the JED and so no need to continue. This is a policy, and only can be changed by a vote in the team.
2. Yes it cost nothing... but if we return as expected then it is dead code, which is wrong code.

Reply:
Keeping the possibility open for future change is okay, I am very open to that change. But lets not code the change in the logic just yet... ;)

Reply:
Templates are not allowed by itself, but they may be installed as a part of a package (e.g. see "Minima Template for SEBLOD", or "Mobile" extension before recent separation of core and template, and probably some of premium "coming soon"-like extensions could do its work via a supplied template). And IMHO in such a case we should apply all JED requirements to the supplied template as well. At the same time, if a template is the extension type in the "root" manifest file, corresponding error will be displayed in the "Info" rules section.

Reply:
Seems Responsivizer extension contains a template too (but I'm not sure as it's paid extension).